src/main/java/dk/cngroup/trainings/spring/springassignment/controller/AnimalController.java

```diff
@@ -14,61 +17,102 @@
 @RequestMapping("/animals")
 public class AnimalController {
 
-    private AnimalService animalService;
+	private AnimalService animalService;
 
-    public AnimalController(AnimalService animalService) {
-        this.animalService = animalService;
-    }
+	public AnimalController(AnimalService animalService) {
+		this.animalService = animalService;
+	}
 
-    @RequestMapping(path = "", method = RequestMethod.POST)
-    public ResponseEntity<Animal> addAnimal(@RequestBody @Valid Animal animal){
-        if(animalService.isValid(animal)){ //TODO duplication
-            return new ResponseEntity<>(animalService.addAnimal(animal).get(),
-                    HttpStatus.OK); //TODO return OK only ifPresent is true
-        }else{
-            return ResponseEntity.badRequest().build();
-        }
-    }
+	@RequestMapping(path = "", method = RequestMethod.POST)
+	public ResponseEntity<Animal> addAnimal(@RequestBody @Valid Animal animal) {
+		try {
+			return new ResponseEntity<>(animalService.addAnimal(animal).get(), HttpStatus.OK);
+		} catch (InvalidAnimalException e) {
+			return ResponseEntity.badRequest().build();
+		}
+	}
 
-    @RequestMapping(path = "", method = RequestMethod.GET)
-    public ResponseEntity<List<Animal>> getAnimals(){
-        return new ResponseEntity<>(animalService.getAnimals(),
-                HttpStatus.OK);
-    }
+	@RequestMapping(path = "/{id}", method = RequestMethod.PUT)
+	public ResponseEntity<Animal> updateAnimalById(@PathVariable("id") long id,
+												   @RequestBody @Valid Animal animal) {
+		if (animalService.getAnimalById(id).isPresent()) {
```

Comment:
you should not have logic here, you should rather use the update method which you already have in your service

Reply:
Fixed